drivers: usb_dc_stm32: added usb_dc_detach code

StefJar opened the pull request:
there is not so much to do if we detaching. At least we can run
'HAL_PCD_DeInit' to deinit the device. That will save some power.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

Commit:
usb: stm32: added usb_dc_detach code

There is not so much to do if we detaching. At least we can run
'HAL_PCD_DeInit' to deinit the device. That will save some power.

Updated source to the latest gpio API.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

Commit:
usb: stm32: added usb_dc_detach code

there is not so much to do if we detaching. At least we can run
'HAL_PCD_DeInit' to deinit the device. That will save some power.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

Commit:
usb: stm32: added usb_dc_detach code

there is not so much to do if we detaching. At least we can run
'HAL_PCD_DeInit' to deinit the device. That will save some power.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

carlescufi commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -990,10 +989,17 @@ void HAL_PCDEx_SetConnectionState(PCD_HandleTypeDef *hpcd, uint8_t state)
+			   DT_INST_0_ST_STM32_USB_DISCONNECT_GPIOS_PIN, GPIO_DIR_OUT);
+
+	if (state) {
+		gpio_pin_write(usb_disconnect,
```
This function is deprecated. See the mailing list for the GPIO announcement, this needs to use the new GPIO API.

StefJar replied:
I only changed "usb_dc_detach". The changes shown at line 993 are only marked because the file got  a bit shorter. ;)

jfischer-phytec-iot replied:
@StefJar new GPIO API was merged recently. Your patch changes to gpio_ping_configure() are wrong/unnecessary, probably failed rebase/conflict resolution.

StefJar replied:
hey ho
merged  a latest zephyr master branch and re-commited that changes. Added a extra commend. So it should be fine for you guys

jfischer-phytec-iot commented:
FYI: #16026 is a stale PR and also tries to implement usb detach functionality for STM32 platform.

Commit:
usb: stm32: added usb_dc_detach code

There is not so much to do if we detaching. At least we can run
'HAL_PCD_DeInit' to deinit the device. That will save some power.

Updated source to the latest gpio API.

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

emob-nordic commented:
ammend the commits please.

StefJar commented:
> ammend the commits please.

done - see the commitment commend ;)

jfischer-phytec-iot commented:
@StefJar Based on master your patch should be like:
 
```diff
diff --git a/drivers/usb/device/usb_dc_stm32.c b/drivers/usb/device/usb_dc_stm32.c
index d9ad95f456..a1676cae7c 100644
--- a/drivers/usb/device/usb_dc_stm32.c
+++ b/drivers/usb/device/usb_dc_stm32.c
@@ -846,7 +846,7 @@ int usb_dc_ep_mps(const u8_t ep)
 
 int usb_dc_detach(void)
 {
-       LOG_ERR("Not implemented");
+       HAL_PCD_DeInit(&usb_dc_stm32_state.pcd);
 
        return 0;
 }

```

Apart from that, does it work to detach and attach?

jfischer-phytec-iot commented:
@erwango Should peripheral clock be disabled, like it was intended in #16026?

StefJar commented:
> @StefJar Based on master your patch should be like:
> 
> ```diff
> diff --git a/drivers/usb/device/usb_dc_stm32.c b/drivers/usb/device/usb_dc_stm32.c
> index d9ad95f456..a1676cae7c 100644
> --- a/drivers/usb/device/usb_dc_stm32.c
> +++ b/drivers/usb/device/usb_dc_stm32.c
> @@ -846,7 +846,7 @@ int usb_dc_ep_mps(const u8_t ep)
>  
>  int usb_dc_detach(void)
>  {
> -       LOG_ERR("Not implemented");
> +       HAL_PCD_DeInit(&usb_dc_stm32_state.pcd);
>  
>         return 0;
>  }
> ```
> 
> Apart from that, does it work to detach and attach?
With my current firmware I can't test it. I like to have a real "usb_disable()" on a stm32f412 in order to save energy from the un-powered usb when my device sleeps. Via LSE powered events I decide if I need to reboot my device.
@periphal clk: the usb periphal has a own deep sleep mode. see RM0402 at 29.2.1 for the sm32f412. We only need to signal it

erwango reviewed (changes requested):
@StefJar can you have another look? there are 2 commits for a one line change.

eobalski reviewed (changes requested):
Switch to Your branch and:
git reset HEAD~1
git add drivers/usb/device/usb_dc_stm32.c
git commit --amend
git push -f upstream master

Commit:
Merge branch 'master' of github.com:StefJar/zephyr

carlescufi commented:
@emob-nordic @jfischer-phytec-iot please revisit

erwango commented:
@StefJar there are 4 commit now.
One thing you can do is:
- `git co origin/master -B usb_detach`  > create a new branch usb_detach from current master of repo origin and check it out
- `git cherry-pick aac044360975e80fe42ce9f6e1effba169eb354b` > apply the commit we're interested in on the new branch
- `git co -B usb_detach master` > create a new local branch 'master' from local branch usb_detach and check it out
- `git push -f origin master ` > force push the local master branch

The 2 last step are required because you used a local branch named `master` to push the current PR and so we're need to force puh from the same branch to update current PR. Ideally you should create  "topic local branch" with a dedicated name for each PR you want to push. It avoid creating confusion with origin/master.

emob-nordic commented:
@StefJar I am okay with the code, however we would like to avoid merging 4 commits into master with one line change. What You can do is squash those commits into one. You can follow [example](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html).
Let me know if You need any help, You can catch me on Slack.

StefJar commented:
agreed
I will try to fix it over the day. I will give @erwango idea a try.
What a mess from only one line of code

erwango commented:
@StefJar there was an issue in one of the commands. I've edited. Sorry for that .

ydamigos reviewed (commented):
@StefJar How did you test the change? Which sample?
In PR #16026, when HAL_PCD_DeInit() was called, the application got stuck. Did you face a similar problem?
@jli157 FYI.

StefJar commented:
That got now super messy. I will close it and resubmitted it
sorry guys for the effort

StefJar commented:
the new pull request is
https://github.com/zephyrproject-rtos/zephyr/pull/22721
